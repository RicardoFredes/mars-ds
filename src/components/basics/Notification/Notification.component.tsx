import { NotificationProps } from "./Notification.types";
import classNames from "classnames";
import Card from "@/components/basics/Card";
import { CardElevations } from "@/components/basics/Card/card.types";
import Icon from "@/components/basics/Icon";
import Tokens from "@/tokens";
import { NotificationVariants } from ".";
import Subtitle from "@/components/typographics/Subtitle";
import { useState } from "react";

const icons: Record<
  typeof NotificationVariants[keyof typeof NotificationVariants],
  { name: string; color: string }
> = {
  warning: {
    name: "alert",
    color: Tokens.ColorWarning500,
  },
  success: {
    name: "checkmark-circle",
    color: Tokens.ColorSuccess500,
  },
  error: {
    name: "info",
    color: Tokens.ColorError500,
  },
  info: {
    name: "info",
    color: Tokens.ColorInfo500,
  },
};

const Notification = ({
  className,
  children,
  variant = NotificationVariants.Default,
  onClose,
  ...props
}: NotificationProps) => {
  const [closing, setClosing] = useState(false);
  const cn = classNames("notification", { "notification-is--closing": closing }, className);

  const icon = icons[variant];

  const handleClose = () => {
    setClosing(true);
    onClose?.();
  };

  return (
    <Card className={cn} {...props} elevation={CardElevations.XXHigh}>
      {icon && (
        <Icon name={icon.name} className="notification__icon" style={{ color: icon.color }} />
      )}
      <Subtitle className="notification__title">{children}</Subtitle>

      <Icon onClick={handleClose} name="close" className="notification__close-icon" />
    </Card>
  );
};

export default Notification;
