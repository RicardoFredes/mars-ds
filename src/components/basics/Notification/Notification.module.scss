.notification {
  display: flex;
  align-items: center;
  padding: $spacing-lg;
  padding-right: 64px;
  border-radius: $border-radius-sm;
  width: fit-content;
  animation: animation-open 0.3s ease-out forwards;

  &__icon {
    margin-right: $spacing-md;
  }

  &__content {
    text-overflow: ellipsis;
    overflow: hidden;
  }

  &__close-button {
    position: absolute;
    right: 4px;
    top: 50%;
    transform: translateY(-50%);
  }

  &-is--closing {
    animation: animation-closing 0.3s ease-out forwards;
  }

  @keyframes animation-closing {
    from {
      opacity: 1;
    }
    to {
      opacity: 0;
      transform: translateX(120%);
    }
  }
  @keyframes animation-open {
    from {
      opacity: 0;
      transform: translateX(120%);
    }
    to {
      opacity: 1;
      transform: translateX(0%);
    }
  }
}

.notification-list {
  position: fixed;
  padding: $spacing-lg;
  top: 0;
  right: 0;
  z-index: $elevation-notification;
  display: flex;
  flex-direction: column-reverse;
  gap: $spacing-lg;
}
