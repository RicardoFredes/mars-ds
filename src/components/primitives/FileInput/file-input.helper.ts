export const FIVE_MB = 1 * 1024 * 1024;

export const isValidExtension = (extension: string, extensions?: string[]) => {
  if (!extensions) return true;

  for (const ext of extensions) {
    const reg = new RegExp(`^${ext}$`, "i");
    if (reg.test(extension)) return true;
  }

  return false;
};

export const dictionary = {
  notifications: {
    notFound: "Nenhum arquivo selecionado",
    notAllowed: "Arquivo inv√°lido. Selecione um arquivo do tipo:",
    tooBig: "Arquivo muito grande. Selecione um arquivo menor que",
  },
};
