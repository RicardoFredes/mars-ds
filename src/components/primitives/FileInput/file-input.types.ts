import type { HTMLProps } from "react";

interface FileInputProps extends HTMLProps<HTMLInputElement> {
  extensions?: string[];
  maxSize?: number;
  placeholderImage?: string;
  notifications?: {
    notFound?: string;
    notAllowed?: string;
    tooBig?: string;
  };
  onSelectFile?: (file: File | null) => void;
  onSelectFileStart?: VoidFunction;
  onSelectFileEnd?: VoidFunction;
  onUploadFile?: (file: string | null) => void;
}

export type { FileInputProps };
