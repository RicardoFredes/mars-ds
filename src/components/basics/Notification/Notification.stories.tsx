import Button from "@/components/basics/Button";
import type { ComponentMeta, ComponentStory } from "@storybook/react";
import Notification from "./";
import { NotificationVariants } from "./Notification.types";

export default {
  title: "Components/Notification",
  component: Notification,
  argTypes: {
    variant: {
      options: NotificationVariants,
      control: { type: "select" },
    },
  },
} as ComponentMeta<typeof Notification>;

const Template: ComponentStory<typeof Notification> = (args) => <Notification {...args} />;

const DefaultArgs = {
  variant: NotificationVariants.Default,
  children: "Mensagem de aviso.",
};

export const Default = Template.bind({});
Default.args = DefaultArgs;

export const Warning = Template.bind({});
Warning.args = {
  ...DefaultArgs,
  variant: NotificationVariants.Warning,
};

export const Success = Template.bind({});
Success.args = {
  ...DefaultArgs,
  variant: NotificationVariants.Success,
};
export const VariantError = Template.bind({});
VariantError.args = {
  ...DefaultArgs,
  variant: NotificationVariants.Error,
};
export const Info = Template.bind({});
Info.args = {
  ...DefaultArgs,
  variant: NotificationVariants.Info,
};

const ButtonTemplate: ComponentStory<typeof Notification> = (args) => (
  <Button onClick={() => Notification.open(args)} label="Abrir Notificação" />
);

export const OpenNotification = ButtonTemplate.bind({});
OpenNotification.args = {
  ...DefaultArgs,
  durationInMs: 5000,
};
