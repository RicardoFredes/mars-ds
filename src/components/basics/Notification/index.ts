/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 */

import Notification from "./notification.component";
export default Notification;
export * from "./notification.types";
