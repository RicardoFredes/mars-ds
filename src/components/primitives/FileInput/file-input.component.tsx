import type { FileInputProps } from "./file-input.types";
import type { ChangeEvent } from "react";

import classNames from "classnames";

import Notification from "@/components/basics/Notification";

import { dictionary as commonDictionary, FIVE_MB, isValidExtension } from "./file-input.helper";

const FileInput = ({
  extensions,
  maxSize = FIVE_MB,
  notifications,
  className,
  onSelectFile,
  onSelectFileStart,
  onSelectFileEnd,
  onUploadFile,
  ...props
}: FileInputProps) => {
  const cn = classNames("file-input", className);

  const errors = { ...commonDictionary.notifications, ...notifications };

  const handleSelectFile = async (file: File) => {
    onSelectFileStart?.();
    onSelectFile?.(file);

    const updatedFile = await new Promise<string>((resolve, reject) => {
      const reader = new FileReader();

      reader.readAsDataURL(file);

      reader.onload = () => {
        const result = reader.result;
        if (result) resolve(result.toString());
        else reject();
      };
    });

    onUploadFile?.(updatedFile);
    onSelectFileEnd?.();
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files?.[0];

    if (!file) {
      Notification.error(errors.notFound);
      return;
    }

    const extension = file.name.replace(/.*\./, "");

    if (!isValidExtension(extension, extensions)) {
      Notification.error(`${errors.notAllowed} ${extensions?.join(" ou ")}`);
      return;
    }

    if (file.size > maxSize) {
      const sizeInMB = Math.round(maxSize / 1024 / 1024);
      Notification.error(`${errors.tooBig} ${sizeInMB}MB`);
      return;
    }

    handleSelectFile(file);
  };

  return (
    <input type="file" className={cn} data-testid="file-input" onChange={handleChange} {...props} />
  );
};

export default FileInput;
