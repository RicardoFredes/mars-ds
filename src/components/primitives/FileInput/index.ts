/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 * run "yarn reindex" to update
 */

import FileInput from "./file-input.component";
export default FileInput;
export * from "./file-input.types";
export * from "./file-input.helper";
