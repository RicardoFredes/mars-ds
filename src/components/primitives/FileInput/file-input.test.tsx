import type { FileInputProps } from "./file-input.types";

import { fireEvent, render, waitFor } from "@testing-library/react";

import Notification from "@/components/basics/Notification";

import FileInput from "./file-input.component";
import { isValidExtension } from "./file-input.helper";

describe("FileInput", () => {
  describe("isValidExtension", () => {
    it("should return true if no extensions are defined", () => {
      expect(isValidExtension("png")).toBe(true);
    });

    it("should return true if the extension is valid", () => {
      expect(isValidExtension("png", ["png"])).toBe(true);
    });

    it("should return false if the extension is not valid", () => {
      expect(isValidExtension("png", ["jpg"])).toBe(false);
    });
  });

  describe("<FileInput>", () => {
    const notificationErrorSpy = jest.spyOn(Notification, "error");
    const inputTestId = "file-input";

    const defaultProps: FileInputProps = {
      extensions: ["jpg", "jpeg", "png"],
    };

    const makeSut = (props?: FileInputProps) => render(<FileInput {...defaultProps} {...props} />);

    it("should not accept files with extensions different than allowed", () => {
      const { getByTestId } = makeSut({ extensions: ["jpg", "jpeg"] });
      const file = new File([""], "file.png", { type: "image/png" });
      const input = getByTestId(inputTestId);
      const event = { target: { files: [file] } };

      fireEvent.change(input, event);
      expect(notificationErrorSpy).toBeCalledWith(
        "Arquivo invÃ¡lido. Selecione um arquivo do tipo: jpg ou jpeg"
      );
    });

    it("should accept files with extensions allowed", async () => {
      const onSelectFileSpy = jest.fn();
      const onUploadFileSpy = jest.fn();
      const wrapper = makeSut({
        extensions: ["jpg", "png"],
        onSelectFile: onSelectFileSpy,
        onUploadFile: onUploadFileSpy,
      });

      const file = new File(["1"], "file.png", { type: "image/png" });
      const input = wrapper.getByTestId(inputTestId);
      const event = { target: { files: [file] } };

      await waitFor(() => {
        fireEvent.change(input, event);
        expect(onSelectFileSpy).toHaveBeenCalledWith(file);
        expect(onUploadFileSpy).toHaveBeenCalledWith("data:image/png;base64,MQ==");
        expect(wrapper.container).toMatchSnapshot();
      });
    });

    it("should not accept files with size greater than maxSize", () => {
      const wrapper = makeSut();

      const input = wrapper.getByTestId(inputTestId);
      const event = {
        target: {
          files: [
            {
              name: "image.png",
              size: 10 * 1024 * 1024,
              type: "image/png",
            },
          ],
        },
      };
      fireEvent.change(input, event);
      expect(notificationErrorSpy).toBeCalledWith(
        "Arquivo muito grande. Selecione um arquivo menor que 1MB"
      );
      expect(wrapper.container).toMatchSnapshot();
    });
  });
});
