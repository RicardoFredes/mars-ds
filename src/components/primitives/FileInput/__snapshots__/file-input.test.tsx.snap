// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput <FileInput> should accept files with extensions allowed 1`] = `
<div>
  <input
    class="file-input"
    data-testid="file-input"
    type="file"
  />
</div>
`;

exports[`FileInput <FileInput> should not accept files with size greater than maxSize 1`] = `
<div>
  <input
    class="file-input"
    data-testid="file-input"
    type="file"
  />
</div>
`;
